argo_workflow_tools/dsl/dag_compiler.py

```diff
@@ -178,6 +178,28 @@ def build_condition(conditions: List[Union[BinaryOp, UnaryOp]]):
     return "&&".join(condition_expr)
 
 
+def _build_exit_hook(exit_hook: Callable) -> Dict[str, argo.LifecycleHook]:
+    if exit_hook:
+        ctx = copy_context()
+        dag_output = ctx.run(exit_hook)
+        dag_tasks = ctx.get(workflow_template_collector.dag_tasks, [])
+        arguments = [
+            _build_node_input(input_name, input_type)
+            for input_name, input_type in dag_tasks[0].arguments.items()
+        ]
+        if isinstance(dag_tasks[0], DAGReference):
+            _build_dag_template(dag_tasks[0])
+        elif isinstance(dag_tasks[0], TaskReference):
+            _build_task_template(dag_tasks[0])
```

Comment:
getting the first task in the dag happens quite a lot, maybe call it once before this logic, and use the ref instead?